### PR TITLE
scheduler: refactor clusterInfo to extract a BasicCluster, and add mockCluster to easy testing

### PR DESCRIPTION
**Pull request by qqsun8819**
This pr resolves issue [issue 752](https://github.com/pingcap/pd/issues/752)
1 add mockCluster in schedulers , and implements basic interface in schedule.Cluster.
2 move balance_test.go into schedulers
3 remove balance.go in server for duplicate code of schedulers/util.go

On the other hand, this pr imports some new bad smell code:
1 mockCluster has some duplicate code of clusterInfo in server/cache.go, due to dependency problem. 
2 coordinator_test.go also some duplicate code of mockCluster, this is a remaining problem after we move balance_test.go out of server. 
I plan to propose a new pr to resolve these two problems after this pr is merged

### CODE REVIEW

File: server/schedulers/mockcluster.go [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+	id              *core.MockIDAllocator
+	stores          *core.StoresInfo
+	regions         *core.RegionsInfo
+	writeStatistics cache.Cache
+	readStatistics  cache.Cache
+}
+
+// NewMockCluster creates a new mockCluster
+func newMockCluster(id *core.MockIDAllocator) *mockCluster {
+	return &mockCluster{
```
Comment (nolouch) [inline server/schedulers/mockcluster.go]: maybe we can let `clusterinfo` be public and use it directly

Comment (qqsun8819) [reply server/schedulers/mockcluster.go]: ok， I'll correct it.

**Comment by qqsun8819** [comment]
@disksing @nolouch We need to discuss about how we deal with clusterInfo in cache.go, as @nolouch proposed in his comment.
As I mentioned in pr description,  now mockCluster has many duplicate code as clusterInfo, so we should extract clusterInfo along into a separate package; but as clusterInfo depends on many data structure in server package ( struct in replication.go etc.) , these module also has to be moved into separate package like core. This change is somewhat big...
According to these problems , I have some proposals we can discuss so we can proceed for further refactor:
1  move clusterInfo into "schedules" package, or into a new package like "cluster".
2 resolve all the dependency problems caused by the movement of clusterInfo in 1.
3  whether 1 and 2  should be done in  a new pr after this pr is merged, or do it just in this pr.
As 1 and 2 caused many changes, what I prefer is open a new pr to do this, and I need your suggestion.

File: server/schedulers/balance_test.go [hunk 2]
```diff
@@ -11,178 +11,24 @@
...
-	"time"
-
 	. "github.com/pingcap/check"
-	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/pingcap/kvproto/pkg/pdpb"
 	"github.com/pingcap/pd/server/cache"
 	"github.com/pingcap/pd/server/core"
 	"github.com/pingcap/pd/server/schedule"
-	_ "github.com/pingcap/pd/server/schedulers" // Register schedulers for tests.
+	"math"
```
Comment (disksing) [inline server/schedulers/balance_test.go]: put it before 3rd-part libraries.

**Comment by disksing** [comment]
How about extract a cluster struct in `core` or `scheduler` contains following four fields:
```
stores          *core.StoresInfo
regions         *core.RegionsInfo
writeStatistics cache.Cache
readStatistics  cache.Cache
```
Then embed it into `clusterInfo` and `mockClusterInfo`.

@qqsun8819

**Comment by siddontang** [comment]
Any update @qqsun8819

**Comment by qqsun8819** [comment]
@siddontang I'm working on it , plan to complete it this weekend according to @disksing 's suggestion

**Comment by qqsun8819** [comment]
@disksing @nolouch @siddontang PR is updated according to review comments, PTAL if you have time

File: server/cache_test.go [hunk 3]
```diff
@@ -337,22 +337,22 @@ func (s *testClusterInfoSuite) testRegionHeartbeat(c *C, cache *clusterInfo) {
 	n, np := uint64(3), uint64(3)
 
 	stores := newTestStores(3)
-	regions := newTestRegions(n, np)
+	Regions := newTestRegions(n, np)
```
Comment (disksing) [inline server/cache_test.go]: Don't change local variable names...

File: server/cache_test.go [hunk 4]
```diff
@@ -84,10 +84,10 @@ var _ = Suite(&testRegionsInfoSuite{})
 
 type testRegionsInfoSuite struct{}
 
-// Create n regions (0..n) of n stores (0..n).
+// Create n Regions (0..n) of n stores (0..n).
```
Comment (disksing) [inline server/cache_test.go]: Don't need to change the comment.

File: server/core/test_util.go [hunk 5]
```diff
@@ -78,3 +79,18 @@ func NewRegion(start, end []byte) *metapb.Region {
...
+type MockIDAllocator struct {
+	base uint64
+}
+
+// NewMockIDAllocator create a new MockIDAllocator
+func NewMockIDAllocator() *MockIDAllocator {
+	return &MockIDAllocator{base: 0}
+}
+
+// Alloc return a new id
```
Comment (disksing) [inline server/core/test_util.go]: Alloc **returns** a new id.

File: server/schedulers/balance_test.go [hunk 6]
```diff
@@ -11,179 +11,25 @@
...
-
 	. "github.com/pingcap/check"
-	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/pingcap/kvproto/pkg/pdpb"
 	"github.com/pingcap/pd/server/cache"
 	"github.com/pingcap/pd/server/core"
 	"github.com/pingcap/pd/server/namespace"
 	"github.com/pingcap/pd/server/schedule"
-	_ "github.com/pingcap/pd/server/schedulers" // Register schedulers for tests.
+	"math"
```
Comment (disksing) [inline server/schedulers/balance_test.go]: We usually put builtin packages before 3rd-part ones.

File: server/cache.go [hunk 7]
```diff
@@ -36,25 +36,19 @@ var (
 )
 
 type clusterInfo struct {
+	*schedule.BasicCluster
 	sync.RWMutex
```
Comment (siddontang) [inline server/cache.go]: The RWMutex should also wrap the basic cluster, if yes, move the RWMutex to the first line of the clusterInfo.

File: server/cache.go [hunk 8]
```diff
@@ -73,16 +67,16 @@ func loadClusterInfo(id core.IDAllocator, kv *core.KV) (*clusterInfo, error) {
 	}
 
 	start := time.Now()
-	if err := kv.LoadStores(c.stores, kvRangeLimit); err != nil {
+	if err := kv.LoadStores(c.Stores, kvRangeLimit); err != nil {
 		return nil, errors.Trace(err)
 	}
-	log.Infof("load %v stores cost %v", c.stores.GetStoreCount(), time.Since(start))
+	log.Infof("load %v Stores cost %v", c.Stores.GetStoreCount(), time.Since(start))
```
Comment (siddontang) [inline server/cache.go]: using stores and regions in the log and comment is ok, no need to change them.

File: server/cache_test.go [hunk 9]
```diff
@@ -271,35 +271,35 @@ func (s *testClusterInfoSuite) TestLoadClusterInfo(c *C) {
...
-		c.Assert(region, DeepEquals, regions[region.GetId()])
+		c.Assert(region, DeepEquals, Regions[region.GetId()])
 	}
 }
 
 func (s *testClusterInfoSuite) testStoreHeartbeat(c *C, cache *clusterInfo) {
 	n, np := uint64(3), uint64(3)
 	stores := newTestStores(n)
-	regions := newTestRegions(n, np)
+	Regions := newTestRegions(n, np)
```
Comment (siddontang) [inline server/cache_test.go]:
```
regions := 
```

File: server/cache.go (hunk 8 above)
Comment (qqsun8819) [reply server/cache.go]: done

File: server/cache_test.go (hunk 4 above)
Comment (qqsun8819) [reply server/cache_test.go]: done

File: server/cache_test.go (hunk 3 above)
Comment (qqsun8819) [reply server/cache_test.go]: done

File: server/schedulers/balance_test.go (hunk 6 above)
Comment (qqsun8819) [reply server/schedulers/balance_test.go]: done

File: server/cache_test.go (hunk 9 above)
Comment (qqsun8819) [reply server/cache_test.go]: done

**Comment by disksing** [comment]
LGTM.

**Pull request merged**